### DISCUSSION TITLE
Better way to get output from patch operations

### DISCUSSION THREAD

**Discussion by ArkAnsen** [ravendb/ravendb#12557] [Q&A]
I've got a patch that updates several leaf documents in a transactional manner (see below) and I would like to use the `output()` of the patch to return log data of the patch operation. 

A couple notes:
- This patch is part of a multi-step courier transaction. I need to keep a record of the old serial number in the event a compensation is required.
- I do not need the entire document to verify that the serial number has indeed changed. For performance reasons I would like to avoid a call and round trip created by `session.Load()`
- For concurrency reasons I would like to avoid a call to `session.Load()` and stick with patching

Is there a better way to do this? 
Here is what I currently have as a test case using LinqPad:
```csharp
const string id = "devices/5514";
const string serialNumber = "TEMP-ABCD1234";

const string patchScript = @"
        output({ OldSerialNumber: this.SerialNumber, NewSerialNumber: args.SerialNumber });
	if (this.SerialNumber !== args.SerialNumber) 
		return;
		
	this.SerialNumber = args.SerialNumber;
	
	var leafDoc1 = load(this.LeafDoc1);
        var leafDoc2 = load(this.LeafDoc2);
        var leafDoc3 = load(this.LeafDoc3);
        var leafDoc4 = load(this.LeafDoc4);
        var leafDoc5 = load(this.LeafDoc5);
        var leafDoc6 = load(this.LeafDoc6);
        var leafDoc7 = load(this.LeafDoc7);
        var leafDoc8 = load(this.LeafDoc8);

        leafDoc1.SerialNumber = args.SerialNumber;
        leafDoc2.SerialNumber = args.SerialNumber;
        leafDoc3.SerialNumber = args.SerialNumber;
        leafDoc4.SerialNumber = args.SerialNumber;
        leafDoc5.SerialNumber = args.SerialNumber;
        leafDoc6.SerialNumber = args.SerialNumber;
        leafDoc7.SerialNumber = args.SerialNumber;
        leafDoc8.SerialNumber = args.SerialNumber;
		
	var location = load(this.Location.Id);
	_.forEach(location.Devices, function(value) {
		if(value.Id !== this.Id) 
			return;
		
		value.SerialNumber = args.SerialNumber;
	});
";

var patchOperation = new PatchOperation(
	id: id,
	changeVector: null,
	patch: new PatchRequest { Script = patchScript, Values = { { "SerialNumber", serialNumber } } }
);
	
using (var session = store.OpenAsyncSession())
{
	var cmd = patchOperation.GetCommand(store, store.Conventions, session.Advanced.Context, null, true, true);
	await session.Advanced.RequestExecutor.ExecuteAsync(cmd, session.Advanced.Context);
	var result = (BlittableJsonReaderObject)cmd.Result.Debug;
	
	using (var stream = new MemoryStream())
	{
		((BlittableJsonReaderObject)cmd.Result.Debug).WriteJsonTo(stream);
		var debugInfo = Encoding.UTF8.GetString(stream.ToArray());
		var debugObject = JsonConvert.DeserializeObject<JObject>(debugInfo);
		var log = JsonConvert.DeserializeObject<UpdateSerialNumberLog>(debugObject["Info"].First().Value<string>());
		$"Serial Number updated from {log.OldSerialNumber} to {log.NewSerialNumber}".Dump();
	}
}
```

**Comment by ayende** [comment]
That is the right way to go about doing this, correct.

**Comment by ArkAnsen** [comment]
Could the client ergonomics of this be improved? The process is quite brittle for what in the end is returning a JSON document / object. The functionality to easily return a JSON document from a patch paves the way for some interesting patching operations.

Still poking at this, the following is cleaner, though still circuitous for getting the intended result. 

```csharp
using (var session = store.OpenAsyncSession())
{
	var cmd = patchOperation.GetCommand(store, store.Conventions, session.Advanced.Context, null, true, true);
	await session.Advanced.RequestExecutor.ExecuteAsync(cmd, session.Advanced.Context);
		
	var result = "No Result parsed";
	if (cmd.StatusCode == System.Net.HttpStatusCode.NotFound)
		result = $"Document {id} does not exist";
	else
	{
		cmd.Result.ModifiedDocument.TryGet<string>("SerialNumber", out var nsn);
		cmd.Result.OriginalDocument.TryGet<string>("SerialNumber", out var osn);
		result = cmd.Result.Status switch
		{
			PatchStatus.Patched => $"Serial Number updated from {osn} to {nsn}",
			PatchStatus.NotModified => "Serial Number not modified",
			PatchStatus.DocumentDoesNotExist => throw new InvalidOperationException($"Document {id} does not exist"), // This should rarely happen due to 404 error above?
			PatchStatus.Skipped => "Not modified due to stale data", // This should not happen due to no change vector above
			PatchStatus.Created => $"Document {id} created", // This should not happen due to no patchIfMissing above
			_ => "Unknown State"
		};
	}
		
	result.Dump();		
}
```

**Reply by ayende** [reply]
Yes, a nicer API would be good to have, I agree.

What would you consider a good approach to handle this?

**Reply by ArkAnsen** [reply]
From a very high level I wonder if the client API could be made nearly as simple as the `session.Load<TEntity>()`.
Something like `Task<PatchCommand.Result<TEntity, TResult>> = session.Patch<TEntity, TResult>(new PatchOperation { ... })` which extends the already existing `PatchCommand.Result<TEntity>`

Without knowing the inner workings of JInt, the return could be handled a couple different ways:
- Use the current `output` function and apply the deserialization logic when a `TResult` is requested
- Create a new `outputResult` function and apply the deserialization logic
- Possibly use the Javascript `return TResult`

**Reply by ayende** [reply]
We can do the `return` option, yes. However, I'm not sure how useful that would be.
For that matter, the `output` is meant to allow you to handle multiple result sets.

Note that we are primarily talking about client side behavior, not server side changes. My thinking is to provide a more ergonomic option from the previous code.

**Reply by ArkAnsen** [reply]
I thought about that. The big problem that would be introduced with using `output` for multiple result sets is that there is no easy way to directly extract a specific which result set. This becomes especially problematic for the client if the `output` is placed inside a conditional statement.

If the debug output must support multiple results sets, then the scope of the design must to be expanded to provide an easy way to test for and retrieve a specific result set. Since we are dealing with JavaScript the simplest format is an object/dictionary. The introduction of an overload that accepts a simple string key (`output(<str_key>, <value>)`) that can be easily referenced by the client would make sense. However wouldn't that require updating the server side implementation?

**Reply by ayende** [reply]
You can do that directly using `output({ 'id': str_key, ...})`, no?

**Reply by ArkAnsen** [reply]
Yes, however what happens in this case?
```javascript
output({id: 'key1', value: 'value1'})
if (condition === true)
  output({id: 'key2', value: 'value1'})
output({id: 'key3', value: 'value1'})
```
On the client side I have no way of knowing which output resides at index 1 of the output array. It would be nice if it could be as simple as `(bool, TResult) = patchResult.TryGetOutput<TResult>("key2")`

**Reply by ayende** [reply]
On the client side, you can just return a `List<T>`, where the script decide what the id is.
Remember, we are just returning the shape that the script gave, we don't want to impose here.
